File: sudokuSplitter.py
import cv2
import numpy as np
import math
import sys
import pytesseract


cellSize = 56
border = 3

def focusGrid(ogimg):
    rx = 500.0/ogimg.shape[0]
    ry = 500.0/ogimg.shape[1]
    r = max([rx,ry])


    ogimg = cv2.resize(ogimg,(0,0),fx=r,fy=r)
    img = cv2.cvtColor(ogimg, cv2.COLOR_BGR2GRAY)
    img = cv2.addWeighted(img, 1.5, np.zeros(img.shape, img.dtype), 0, 0)

    img = cv2.adaptiveThreshold(img,255,cv2.ADAPTIVE_THRESH_MEAN_C,
            cv2.THRESH_BINARY,25,25)
    blur = cv2.GaussianBlur(img,(3,3),3)
    
    edged = cv2.Canny(blur,100,180)

    # cv2.imshow("img",img)
    # cv2.imshow("edged",edged)
    # cv2.waitKey(0)
    
    im2, contours, hierarchy = cv2.findContours(edged, cv2.RETR_LIST, cv2.CHAIN_APPROX_SIMPLE)
    
    if len(contours) == 0:
        print("No contours found")
        return None

    cnt = None
    maxArea = 0
    # print(len(contours))
    for c in contours:
        area = cv2.contourArea(c)
        if area > maxArea:
            maxArea = area
            cnt = c

    
    if cnt is None:
        print("No biggest contour")
        return None

    # temp = img

    # contours = sorted(contours,key= lambda c: -cv2.contourArea(c))
    # cv2.drawContours(ogimg, contours[0:1000], -1, (0, 255, 0), 1) 
    # cv2.drawContours(ogimg, [cnt], -1, (0, 255, 0), 1) 
    # cv2.imshow("wow",ogimg)
    # cv2.waitKey(0)


    epsilon = 0.01*cv2.arcLength(cnt,True)
    approx = cv2.approxPolyDP(cnt,epsilon,True)


    if (approx.size != 8):
        print("Wrong shape of grid")
        return
    approx = approx.reshape(4,2)
    approx = rearrangeCorners(approx,ogimg.shape[0],ogimg.shape[1])
    # print(approx)
    approx = np.array(approx.tolist(),np.float32)
    

    gridSize = cellSize*9
    final = np.array([
		[0, 0],
		[0, gridSize],
		[gridSize, gridSize],
		[gridSize, 0]], dtype = "float32")

    M = cv2.getPerspectiveTransform(approx, final)
    fixed = cv2.warpPerspective(img, M, (gridSize, gridSize))

    return fixed

def splitUp(grid):
    cells = []
    for i in range(0,9):
        row = []
        for j in range(0,9):
            cropped = grid[cellSize*i+border:cellSize*(i+1)-border, cellSize*j+border:cellSize*(j+1)-border]
            row.append(cropped)
        cells.append(row)
    return cells

def saveCells(cells):
    for i in range(len(cells)):
        for j in range(len(cells[i])):
            if cells[i][j] is not None:
                cv2.imwrite("digits/grid_"+str(i)+str(j)+".png",cells[i][j])

def distSquared(p1,p2):
    return (p1[0]-p2[0])**2 + (p1[1]-p2[1])**2

def rearrangeCorners(corners,width,height):
    corners = sorted(corners,key=lambda p: distSquared(p,[0,0]))
    tl = corners[0]
    corners = sorted(corners[1:],key=lambda p: distSquared(p,[0,height]))
    bl = corners[0]
    corners = sorted(corners[1:],key=lambda p: distSquared(p,[width,height]))
    return np.array([tl,bl]+corners)

def removeBlackBorder(img,tol=0):
    mask = img>tol
    return img[np.ix_(mask.any(1),mask.any(0))]

def highlightDigit(cell):
    if cell is None:
        return None

    img = cv2.cvtColor(cell,cv2.COLOR_GRAY2RGB)
    # gray = cv2.cvtColor(cell, cv2.COLOR_BGR2GRAY)
    gray = cv2.fastNlMeansDenoising(cell,None,50,5,5)
    gray = cv2.bitwise_not(gray)
    gray = cv2.threshold(gray, 127, 255, cv2.THRESH_BINARY)[1]

    # cv2.imshow("gray",gray)
    # cv2.waitKey(0)

    edges = cv2.Canny(gray,1,1)

    output = cv2.connectedComponentsWithStats(gray, 8, cv2.CV_32S)
    labels = output[1]
    stats = output[2]
    sizes = stats[1:, -1]
    if len(output[2]) <= 1:
        # print("Huh")
        return None
    
    largest_label = 1 + np.argmax(output[2][1:, -1])
    # print(output[2])

    width,height = gray.shape[:2]


    x,y,w,h,_ = stats[largest_label]
    bX = x + w/2.0
    bY = y + h/2.0

    cX = width/2.0
    cY = height/2.0

    tX = cX - bX
    tY = cY - bY

    if (abs(tX) + abs(tY) > 15) or (w*h > 0.9 * width * height):
        # print("No central blob")
        return None

    img[labels != largest_label] = [255,255,255]

    img = img[y:y+h,x:x+w]
    return img

def highlightCells(cells):
    for i in range(len(cells)):
        for j in range(len(cells[i])):
            # print("Trying",i,j)
            cells[i][j] =  highlightDigit(cells[i][j])
            # print(cells[i][j] is not None)
            # if newCell is not None:
            #     cells[i][j] = newCell
    return cells

def thresh(x):
    if x < 2:
        return 0
    if x < 5:
        return 0
    return 1

def printLocations(cells):
    for i in range(len(cells)):
        row = []
        for j in range(len(cells)):
            if cells[i][j] is not None:
                row.append("X")
            else:
                row.append("-")
        print(" ".join(row))

def printCells(cells):
    for i in range(len(cells)):
        row = []
        # print(len(cells[i]))
        for j in range(len(cells[i])):
            if cells[i][j] != 0:
                row.append(str(cells[i][j]))
            else:
                row.append("?")
        # print(len(row))
        print(" ".join(row))

def addPadding(img,border=10):
    return cv2.copyMakeBorder(img,10,10,2,2,cv2.BORDER_CONSTANT,value=[255,255,255])

def hconcat_resize_min(im_list, interpolation=cv2.INTER_CUBIC):
    h_min = min(im.shape[0] for im in im_list)
    im_list = [addPadding(im) for im in im_list]
    im_list_resize = [cv2.resize(im, (int(im.shape[1] * h_min / im.shape[0]), h_min), interpolation=interpolation)
                      for im in im_list]
    return cv2.hconcat(im_list_resize)
    

def flatten(a):
    temp = []
    for row in a:
        temp = temp + row
    return temp

def assertShape(grid,rows=9,cols=9):
    assert(grid is not None)
    assert(len(grid) == rows)
    for row in grid:
        assert(len(row) == cols)
    print("Passed shape test")

def getDigits(cells):
    line = flatten(cells)
    cellsWithDigits = list(filter(lambda x: x is not None, line))
    line = hconcat_resize_min(cellsWithDigits)
    # cv2.imshow("first",line)
    # cv2.waitKey(0)
    # cv2.imwrite("first.png",line)

    custom_config = r'--psm 6 outputbase digits'
    text = pytesseract.image_to_string(line, config=custom_config)

    # print(text)

    if len(text) == 0:
        return None
    
    text = text.partition('\n')
    if len(text) == 0:
        return None

    text = text[0].replace(" ","")
    if len(text) != len(cellsWithDigits) or not text.isdigit():
        return None

    grid = []
    c = 0
    for i in range(0,len(cells)):
        row = []
        for j in range(0, len(cells[i])):
            if cells[i][j] is not None:
                row.append(int(text[c]))
                c += 1
            else:
                row.append(0)
        grid.append(row)
    
    return grid


def extractGrid(img):
    if img is None:
        print("No such image found")
        return None

    clean = focusGrid(img)
    if clean is None:
        print("Failed")
        return None

    cv2.imshow("clean",clean)
    cv2.waitKey(0)

    cells = splitUp(clean)

    cells = highlightCells(cells)

    # printLocations(cells)

    # assertShape(cells)

    # print(cells)
    # saveCells(cells)
    grid = getDigits(cells)

    # assertShape(grid)
    # print(repr(text))
    
    # saveCells(cells)
    # cells = readDigits(cells)
    if grid is None:
        print("Unable to read numbers")
        return None
    
    
    # printCells(grid)

    return grid


    # print("Done")
    # cv2.imshow("Clean",cleas

if __name__ == "__main__":
    grid = focusGrid(cv2.imread(sys.argv[1]))
    cv2.imshow("Grid",grid)
    cv2.waitKey(0)
    # extractGrid(sys.argv[1])

File: sudokuSplitter.js
let cellSize = 56
let border = 3

function focusGrid(img){
    // let img = cv.imread(imgInput)
    // console.log(gray.cols)
    // let graySize

    let gridSize = cellSize*9

    let rx = gridSize/img.cols
    let ry = gridSize/img.rows
    let r = Math.max(rx,ry)


    cv.resize(img,img,new cv.Size(),r,r)

    let gray = new cv.Mat()
    cv.cvtColor(img,gray,cv.COLOR_RGB2GRAY)
    cv.adaptiveThreshold(gray,gray,255,cv.ADAPTIVE_THRESH_MEAN_C,cv.THRESH_BINARY,25,25)
    
    let blur = new cv.Mat()
    cv.GaussianBlur(gray,blur,new cv.Size(3,3),3)
    let edged = new cv.Mat()

    cv.Canny(blur,edged,100,180)

    let contours = new cv.MatVector();
    let hierarchy = new cv.Mat();

    cv.findContours(edged, contours, hierarchy, cv.RETR_LIST, cv.CHAIN_APPROX_SIMPLE);

    if (contours.size == 0){
        console.log("No contours found")
        return null
    }

    // console.log(contours)

    let cntIndex = null
    let maxArea = 0
    for (let i = 0; i < contours.size(); i++){
        let area = cv.contourArea(contours.get(i))
        if (area > maxArea){
            maxArea = area
            cntIndex = i
        }
    }

    if (cntIndex == null){
        console.log("No contours of non-zero area")
        return null
    }

    let epsilon = 0.01*cv.arcLength(contours.get(cntIndex),true)
    let approx = new cv.Mat();
    cv.approxPolyDP(contours.get(cntIndex),approx,epsilon,true)
    // console.log(approx)
    approx = approx.data32S

    // cv.drawContours(img, contours, cntIndex, new cv.Scalar(255,0,0) , 1, cv.LINE_8, hierarchy, 100);

    // console.log(approx)

    if (approx.length != 8){
        console.log("No grid shape found")
        return null
    }

    let srcTri = cv.matFromArray(4, 1, cv.CV_32FC2, approx);
    let dstTri = cv.matFromArray(4, 1, cv.CV_32FC2, [0, 0, 0, gridSize, gridSize, gridSize, gridSize, 0]);
    let M = cv.getPerspectiveTransform(srcTri, dstTri);
    let dsize = new cv.Size(gray.rows, gray.cols);
    cv.warpPerspective(gray, gray, M, dsize, cv.INTER_LINEAR, cv.BORDER_CONSTANT, new cv.Scalar());

    // gray.delete()
    blur.delete()
    edged.delete()

    return gray
}

function splitUp(gridPic){
    // console.log(gridPic.rows)
    // console.log(gridPic.cols)
    let cells = []
    for (let i = 0; i < 9; i++){
        let row = []
        for (let j = 0; j < 9; j++){
            let cropped = new cv.Mat();
            let rect = new cv.Rect(cellSize*j+border,cellSize*i+border, cellSize-2*border,cellSize-2*border)
            // console.log(rect)
            cropped = gridPic.roi(rect);
            row.push(cropped)
        }
        cells.push(row)
    }
    return cells
}

function highlightDigit(cellPic){
    if (cellPic == null){
        return null
    }

    let img = new cv.Mat()
    cv.cvtColor(cellPic,img,cv.COLOR_GRAY2RGB)
    let gray = new cv.Mat()
    // cv.fastNlMeansDenoising(cellPic,gray,null,50,5,5)
    cv.bitwise_not(cellPic,gray)
    let stats = new cv.Mat()
    let labels = new cv.Mat()
    let centroid = new cv.Mat()
    cv.connectedComponentsWithStats(gray, labels, stats, centroid, 8, cv.CV_32S)
    // for (let i = 0; i < stats.rows; i++){
    //     console.log(stats.data32S[i*stats.rows + cv.CC_STAT_AREA])
    // }
    let largestLabel = -1
    let largestArea = -1
    for (let i = 1; i < stats.rows; i++){
        // console.log([stats.data32S[i*stats.rows + cv.CC_STAT_AREA],i])
        let area = stats.data32S[i*stats.cols + cv.CC_STAT_AREA]
        if (area > largestArea){
            largestArea = area
            largestLabel = i
        }
    }

    // console.log(labels)

    let width = cellPic.cols
    let height = cellPic.rows

    let cX = width/2.0
    let cY = height/2.0

    // console.log(Array.from(stats.data32S))
    // console.log([largestLabel,stats.rows])

    let largestInfo = Array.from(stats.data32S).splice(largestLabel*stats.cols,5)


    let x = largestInfo[0]
    let y = largestInfo[1]
    let w = largestInfo[2]
    let h = largestInfo[3]

    // console.log(largestInfo)

    let bX = x + w/2.0
    let bY = y + h/2.0

    if ((Math.abs(bX-cX) + Math.abs(bY-cY)) > 10 || w*h > 0.5*width*height){
        // console.log("No central blob")
        return null
    }

    
    let cropped = new cv.Mat();
    let rect = new cv.Rect(x,y,w,h)
    // console.log(rect)
    cropped = cellPic.roi(rect);

    return cropped
}

function highlightDigits(cellPics){
    for (let i = 0; i < cellPics.length; i++){
        for (let j = 0; j < cellPics[i].length; j++){
            cellPics[i][j] = highlightDigit(cellPics[i][j])
        }
    }
    return cellPics
}

function addPadding(img){
    cv.copyMakeBorder(img,img,10,10,2,2,cv.BORDER_CONSTANT,new cv.Scalar(255))
    return img
}

async function lineUp(imgList){
    minH = Math.min(...imgList.map(x => x.rows))
    imgList = imgList.map(x => addPadding(x))
    // imgList = imgList.map(x => cv.resize(x,x,new cv.Size(),Math.round(x.cols * minH / x.rows),minH))
    for (i in imgList){
        cv.resize(imgList[i],imgList[i],new cv.Size(Math.round(imgList[i].cols * minH / imgList[i].rows),minH))
    }

    // for (img of imgList){
    //     console.log(img.rows)
    // }

    let line = new cv.MatVector()
    for (img of imgList){
        line.push_back(img)
    }

    let compiled = new cv.Mat()
    cv.hconcat(line,compiled)

    cv.cvtColor (compiled, compiled, cv.COLOR_GRAY2RGB)
    return compiled
}

async function getDigits(cells){
    let cellsWithDigits = [].concat.apply([], cells);
    cellsWithDigits = cellsWithDigits.filter(x => x != null)
    let line = await lineUp(cellsWithDigits)
    // publicLine = line
    cv.imshow("output",line)
    line = document.getElementById("output").toDataURL();
    let worker = Tesseract.createWorker({
        logger: m => {}
    });
    let digits = await work(worker,line)
    // console.log(digits)

    if (digits == null || digits.length == 0){
        return null
    }

    digits = digits.split("\n")[0]
    if (digits.length == 0){
        return null
    }

    if ((digits.length != cellsWithDigits.length) || !(/^\d+$/.test(digits))){
        return null
    }

    // console.log(digits)

    let grid = []
    let c = 0
    for (let i = 0; i < cells.length; i++){
        let row = []
        for (let j = 0; j < cells.length; j++){
            if (cells[i][j] != null){
                row.push(parseInt(digits[c], 10))
                c += 1
            } else {
                row.push(0)
            }
        }
        grid.push(row)
    }
    return grid
}


const worker = Tesseract.createWorker({
logger: m => {}
});
    

async function work(worker,img) {
    // console.log(img)

    await worker.load();
    await worker.loadLanguage('eng');
    await worker.initialize('eng');

    await worker.setParameters({
        tessedit_char_whitelist: '0123456789',
        user_defined_dpi: '70'
    });

    let result = await worker.detect(img);
    // console.log(result.data);

    result = await worker.recognize(img);
    // console.log(result.data);
    

    await worker.terminate();

    // let digits = result.data

    return result.data.text
}

let publicCells = []
let publicLine = null

async function solvePic(img){
    if (img == null){
        console.log("No image")
        return false
    }

    let clean = focusGrid(img)
    if (clean == null){
        console.log("Failed to find grid")
        return false
    }

    let cells = splitUp(clean)
    publicCells = cells
    cells = highlightDigits(cells)
    
    let grid = await getDigits(cells)
    // console.log(grid)

    let mainGrid = new Grid(9,3,3)
    await mainGrid.load(grid)

    let result  = await mainGrid.solve()

    console.log(mainGrid.getGrid())
    // console.log(cells)

    // console.log(cells)

    // cv.imshow("output",clean)


}